packages/vscode-extension/src/debug/taskTerminal/devTunnelTaskTerminal.ts

```diff
@@ -0,0 +1,290 @@
+    super(taskDefinition, 1);
+    this.args = taskDefinition.args as IDevTunnelArgs;
+    this.isOutputSummary = false;
+    this.tunnelManagementClientImpl = new TunnelManagementHttpClient(
+      TunnelManagementUserAgent,
+      async () => {
+        const tokenRes = await tools.tokenProvider.m365TokenProvider.getAccessToken({
+          scopes: DevTunnelScopes,
+          showDialog: true,
+        });
+
+        if (tokenRes.isErr()) {
+          return null;
+        }
+        const res = `Bearer ${tokenRes.value}`;
+        return res;
+      }
+    );
+  }
+
+  async stop(error?: any): Promise<void> {
+    if (DevTunnelTaskTerminal.tunnelTaskTerminals.has(this.taskTerminalId)) {
+      DevTunnelTaskTerminal.tunnelTaskTerminals.delete(this.taskTerminalId);
+      if (!this.isOutputSummary) {
+        this.isOutputSummary = true;
+        await this.outputFailureSummary(devTunnelDisplayMessages, error);
+      }
+      if (this.tunnel) {
+        const deleteTunnel = this.tunnel;
+        this.tunnel = undefined;
+        await this.tunnelManagementClientImpl.deleteTunnel(deleteTunnel);
+      }
+
+      if (this.cancel) {
+        this.cancel();
+      }
+      super.stop(error);
+    }
+  }
+
+  do(): Promise<Result<Void, FxError>> {
+    return Correlator.runWithId(getLocalDebugSession().id, () =>
+      localTelemetryReporter.runWithTelemetryProperties(
+        TelemetryEvent.DebugStartLocalTunnelTask,
+        {
+          // TODO: add dev tunnel telemetry
+          // [TelemetryProperty.DebugTaskId]: this.taskTerminalId,
+          // [TelemetryProperty.DebugTaskArgs]: this.generateTaskArgsTelemetry(),
+        },
+        () => this._do()
+      )
+    );
+  }
+
+  private async _do(): Promise<Result<Void, FxError>> {
+    await this.outputStartMessage(devTunnelDisplayMessages);
+    await this.outputStartDevTunnelStepMessage(devTunnelDisplayMessages);
+    await this.resolveArgs(this.args);
+    // TODO: delete the last debug tunnel if it is not deleted
+    const res = await this.start();
+    if (res.isOk()) {
+      await new Promise<void>((resolve) => {
+        const timeout = setTimeout(() => {
+          resolve();
+        }, DevTunnelTimeout);
+        this.cancel = () => {
+          clearTimeout(timeout);
+          resolve();
+        };
+      });
+    }
+    return res;
+  }
+
+  private async start(): Promise<Result<Void, FxError>> {
+    try {
+      const tunnelAccessControlEntry: TunnelAccessControlEntry = {
+        type:
+          this.args.access === Access.public
+            ? TunnelAccessControlEntryType.Anonymous
+            : TunnelAccessControlEntryType.None,
+        subjects: [],
+        scopes: ["connect"],
+      };
+
+      const tunnel: Tunnel = {
+        ports: [{ portNumber: this.args.port, protocol: this.args.protocol }],
+        accessControl: {
+          entries: [tunnelAccessControlEntry],
+        },
+        tags: [DevTunnelTag],
```

Comment:
Is this tag for teams toolkit internal use or for developer to manage tunnel more easily?
If the second one, do we need to print this to terminal or output to let users know it?

Reply:
Maybe in later pr

Reply:
It is for TeamsToolkit internal use. If a tunnel is created by TeamsToolkit and not closed correctly, it will be deleted when the project F5 again.

Reply:
> Is this tag for teams toolkit internal use or for developer to manage tunnel more easily? If the second one, do we need to print this to terminal or output to let users know it?

Good idea, we can added it to the terminal output. All the output messages will be in another PR and reviewed by PMs.